fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Headers.kt

```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * Returns the number of key/value pairs in the map.
+     */
+    override val size: Int
+        get() = contents.size
+
+    /**
+     * Returns `true` if the map contains the specified [key].
+     */
+    override fun containsKey(key: String): Boolean {
+        return contents.containsKey(HeaderName(key))
+    }
+
+        /**
+     * Returns `true` if the map maps one or more keys to the specified [value].
+     */
+    override fun containsValue(value: HeaderValues): Boolean {
+        return contents.containsValue(value)
+    }
+
+    /**
+     * Returns the value corresponding to the given [key], or `null` if such a key is not present in the map.
+     */
+    override fun get(key: String): HeaderValues {
+        return contents[HeaderName(key)].orEmpty()
+    }
+
+    operator fun set(key: String, value: String) = replace(key, value)
+
+    operator fun set(key: String, values: HeaderValues): Headers {
+        put(key, values)
+        return this
+    }
+
+    /**
+     * Returns `true` if the map is empty (contains no elements), `false` otherwise.
+     */
+    override fun isEmpty(): Boolean {
+        return contents.isEmpty()
+    }
+
+    fun transformIterate(set: (key: String, value: String) -> Any, add: (key: String, value: String) -> Any = set) {
+        for ((key, values) in this) {
+            val header = HeaderName(key)
+            when (Headers.isCollapsible(header)) {
+                true ->
+                    // NOTE: HTTP requires all request properties which can
+                    //     * legally have multiple instances with the same key
+                    //     * to use a comma-separated list syntax which enables multiple
+                    //     * properties to be appended into a single property.
+                    set(key, Headers.collapse(header, values))
+                false -> when (Headers.isSingleValue(header)) {
+                    true -> values.lastOrNull()?.let { set(key, it) }
+                    false -> values.forEach { add(key, it) }
+                }
+            }
+        }
+    }
+
+    override fun toString(): String {
+        return contents.toString()
+    }
+
+    companion object {
+        fun isCollapsible(key: HeaderName): Boolean {
+            return when (key) {
+                // These headers, per RFC, SHOULD NOT be collapsed into a single value
+                HeaderName(SET_COOKIE) -> false
+                else -> true
+            }
+        }
+
+        fun isSingleValue(key: HeaderName): Boolean {
+            return when (key) {
+                // These headers, per RFC, SHOULD only appear once.
+                HeaderName(AGE) -> true
+                HeaderName(CONTENT_TYPE) -> true
+                HeaderName(CONTENT_LENGTH) -> true
+                HeaderName(CONTENT_LOCATION) -> true
+                HeaderName(EXPECT) -> true
+                HeaderName(EXPIRES) -> true
+                HeaderName(LOCATION) -> true
```

Comment:
would you consider having a collection of header names called `singleValueHearders` and chaining this to

`return singleValueHeaders.contains(key)`? 

The above collection feels like a magic variable and by naming the collection it explains what makes a single value.

Reply:
It would be a constant then: `SINGLE_VALUE_HEADERS = listOf(...)` which has a `O(n)`. I don't care for either. This just organically grew.

Reply:
the above implementation is O(n) the only thing that would change would be the syntax

Reply:
I don't see this list is getting anywhere near something that o(n) can be significantly slow though. Just aim for better readability.